arwen/host/execution.go

```diff
@@ -848,39 +872,108 @@ func (host *vmHost) callSCMethod() error {
 	log.Trace("callSCMethod")
 
 	runtime := host.Runtime()
-	vmInput := runtime.GetVMInput()
+	callType := runtime.GetVMInput().CallType
+
+	var err error
+	switch callType {
+	case vm.DirectCall:
+		err = host.callSCMethodDirectCall()
+	case vm.AsynchronousCall:
+		err = host.callSCMethodAsynchronousCall()
+	case vm.AsynchronousCallBack:
+		err = host.callSCMethodAsynchronousCallBack()
+	default:
+		err = arwen.ErrUnknownCallType
+	}
+
+	if err != nil {
+		log.Trace("call SC method failed", "error", err)
+	}
+
+	return err
+}
+
+func (host *vmHost) callSCMethodDirectCall() error {
+	_, err := host.callFunctionAndExecuteAsync()
+	return err
+}
+
+func (host *vmHost) callSCMethodAsynchronousCall() error {
+	isCallComplete, err := host.callFunctionAndExecuteAsync()
+	if !isCallComplete /*|| err != nil*/ {
+		return err
+	}
+
+	async := host.Async()
+	output := host.Output()
+	err = async.SendCrossShardCallback(output.ReturnCode(), output.ReturnData(), output.ReturnMessage())
+
+	return err
+}
+
+func (host *vmHost) callSCMethodAsynchronousCallBack() error {
+	runtime := host.Runtime()
 	async := host.Async()
-	callType := vmInput.CallType
 
-	if callType == vm.AsynchronousCallBack {
-		async.Load()
-		asyncCall, err := async.UpdateCurrentCallStatus()
-		if err != nil {
-			log.Trace("UpdateCurrentCallStatus failed", "error", err)
-			err = async.PostprocessCrossShardCallback()
-			if err != nil {
-				log.Trace("call SC method failed", "error", err)
-			}
-			return err
-		}
+	callerCallID := async.GetCallerCallID()
 
-		runtime.SetCustomCallFunction(asyncCall.GetCallbackName())
+	asyncCall, err := async.UpdateCurrentAsyncCallStatus(
+		runtime.GetSCAddress(),
+		callerCallID,
+		runtime.GetVMInput())
+	if err != nil {
+		log.Trace("UpdateCurrentCallStatus failed", "error", err)
+		return err
+	}
+
+	runtime.SetCustomCallFunction(asyncCall.GetCallbackName())
+
+	isCallComplete, callbackErr := host.callFunctionAndExecuteAsync()
+	if callbackErr != nil {
+		metering := host.Metering()
+		metering.UseGas(metering.GasLeft())
+	}
+
+	// TODO matei-p R2 Returning an error here will cause the VMOutput to be
+	// empty (due to CreateVMOutputInCaseOfError()). But in release 2 of
+	// Promises, CreateVMOutputInCaseOfError() should still contain storage
+	// deletions caused by AsyncContext cleanup, even if callbackErr != nil and
```

Comment:
so storage is not deleted ? I think it is deleted by default as vmOutput will not contain the storage changes in case of error.

Reply:
Storage changes are always lost in case of error, indeed. This is perfectly fine for the current state of `AsyncContext` (the first release, R1), because we will only allow a single level of async calls.

However, later when we enable multi-level async calls (R2), then the VM will have to tell the SCProcessor to delete some specific storage keys even if the `VMOutput` was with error. This is a topic for future discussions.